Mathlib/SetTheory/ZFC/Ordinal.lean

```diff
@@ -230,10 +234,155 @@ theorem _root_.ZFSet.isOrdinal_iff_isWellOrder : x.IsOrdinal ↔
+  rw [toPSet, PSet.mem_def]
+  simpa using ((enumIsoToType o).exists_congr_left (p := fun y ↦ x.Equiv y.1.toPSet)).symm
+
+@[simp]
+theorem rank_toPSet (o : Ordinal) : o.toPSet.rank = o := by
+  rw [toPSet, PSet.rank]
+  conv_rhs => rw [← iSup_succ o]
+  convert (enumIsoToType o).symm.iSup_comp (g := fun x ↦ Order.succ x.1.toPSet.rank)
+  rw [rank_toPSet]
+termination_by o
+decreasing_by rename_i x; exact x.2
+
+/-- The von Neumann ordinal corresponding to a given `Ordinal`, as a `ZFSet`.
+
+The elements of `o.toZFSet` are all `a.toZFSet` with `a < o`. -/
+noncomputable def toZFSet (o : Ordinal.{u}) : ZFSet.{u} :=
+  .mk o.toPSet
+
+@[simp]
+theorem mk_toPSet (o : Ordinal) : .mk o.toPSet = o.toZFSet :=
+  rfl
+
+theorem mem_toZFSet_iff {o : Ordinal} {x : ZFSet} : x ∈ o.toZFSet ↔ ∃ a < o, a.toZFSet = x := by
+  refine Quotient.inductionOn x fun x ↦ ?_
+  rw [toZFSet, mk_eq, ZFSet.mk_mem_iff, mem_toPSet_iff]
+  convert Iff.rfl
+  rw [toZFSet, eq, PSet.Equiv.comm]
+
+@[simp]
+theorem rank_toZFSet (o : Ordinal) : o.toZFSet.rank = o :=
+  rank_toPSet o
+
+@[simp]
+theorem toZFSet_toSet {o : Ordinal} : o.toZFSet.toSet = toZFSet '' Iio o := by
+  ext
+  simp [mem_toZFSet_iff]
+
+private theorem toZFSet_mem_toZFSet_of_lt {a b : Ordinal} (h : a < b) :
+    a.toZFSet ∈ b.toZFSet := by
+  rw [mem_toZFSet_iff]
+  exact ⟨a, h, rfl⟩
+
+private theorem toZFSet_subset_toZFSet_of_le {a b : Ordinal} (h : a ≤ b) :
+    a.toZFSet ⊆ b.toZFSet := by
+  intro x hx
+  obtain ⟨c, hc, rfl⟩ := mem_toZFSet_iff.1 hx
+  exact toZFSet_mem_toZFSet_of_lt (hc.trans_le h)
+
+@[simp]
+theorem toZFSet_mem_toZFSet_iff {a b : Ordinal} : a.toZFSet ∈ b.toZFSet ↔ a < b := by
+  refine ⟨?_, toZFSet_mem_toZFSet_of_lt⟩
+  contrapose!
+  exact fun h ↦ notMem_of_subset (toZFSet_subset_toZFSet_of_le h)
+
+@[simp]
+theorem toZFSet_subset_toZFSet_iff {a b : Ordinal} : a.toZFSet ⊆ b.toZFSet ↔ a ≤ b := by
+  refine ⟨?_, toZFSet_subset_toZFSet_of_le⟩
+  contrapose!
+  exact fun h ↦ not_subset_of_mem (toZFSet_mem_toZFSet_of_lt h)
+
+end Ordinal
+
+namespace ZFSet
+open Ordinal
+
+theorem isOrdinal_toZFSet (o : Ordinal) : IsOrdinal o.toZFSet := by
+  refine ⟨fun x hx y hy ↦ ?_, @fun z y x hz hy hx ↦ ?_⟩
+  all_goals
+    obtain ⟨a, ha, rfl⟩ := mem_toZFSet_iff.1 hx
+    obtain ⟨b, hb, rfl⟩ := mem_toZFSet_iff.1 hy
+  · exact toZFSet_mem_toZFSet_iff.2 (hb.trans ha)
+  · obtain ⟨c, hc, rfl⟩ := mem_toZFSet_iff.1 hz
+    exact toZFSet_mem_toZFSet_iff.2 (hc.trans hb)
+
+theorem IsOrdinal.toZFSet_rank_eq {x : ZFSet} (hx : IsOrdinal x) : x.rank.toZFSet = x :=
+  (IsOrdinal.rank_inj (isOrdinal_toZFSet _) hx).1 (rank_toZFSet _)
+
+theorem isOrdinal_iff_mem_range_toZFSet {x : ZFSet.{u}} :
+    IsOrdinal x ↔ x ∈ Set.range toZFSet.{u} := by
+  refine ⟨fun h ↦ ?_, ?_⟩
+  · rw [← h.toZFSet_rank_eq]
+    exact Set.mem_range_self _
+  · rintro ⟨a, rfl⟩
+    exact isOrdinal_toZFSet a
+
+@[simp]
+theorem toZFSet_zero : toZFSet 0 = ∅ := by
+  ext; simp [mem_toZFSet_iff]
+
+@[simp]
```

Comment:
simpNF error here

Reply:
Ugh, I really need to get rid of the `o + 1 = succ o` simp lemma.